File: examples/python/multiple_streams.py
import mcapable

from common import write_sample_file


def main() -> None:
    path = write_sample_file()
    reader = mcapable.Reader.from_path(str(path))

    print("Chunk analysis:")
    chunk_count = 0
    total_uncompressed = 0
    stream = reader.chunks()
    for chunk in stream:
        chunk_count += 1
        total_uncompressed += chunk.uncompressed_size
    print(f"  {chunk_count} chunks")
    print(f"  {total_uncompressed} bytes uncompressed total")
    reader = stream.into_reader()

    print("\nMessage counts per channel:")
    counts = {}
    stream = reader.messages()
    for msg in stream:
        counts[msg.channel_id] = counts.get(msg.channel_id, 0) + 1
    for channel_id, count in counts.items():
        print(f"  channel {channel_id}: {count} messages")
    reader = stream.into_reader()

    print("\nTime range analysis:")
    min_time = None
    max_time = None
    stream = reader.raw_messages()
    for msg in stream:
        min_time = msg.log_time if min_time is None else min(min_time, msg.log_time)
        max_time = msg.log_time if max_time is None else max(max_time, msg.log_time)
    if min_time is not None and max_time is not None:
        print(f"  start: {min_time}")
        print(f"  end:   {max_time}")
        print(f"  duration: {max_time - min_time}")
    reader = stream.into_reader()

    print("\nStream type comparison:")
    stream = reader.records()
    record_count = sum(1 for _ in stream)
    print(f"  Records: {record_count}")
    reader = stream.into_reader()

    stream = reader.chunks()
    chunk_count = sum(1 for _ in stream)
    print(f"  Chunks: {chunk_count}")
    reader = stream.into_reader()

    stream = reader.raw_messages()
    raw_count = sum(1 for _ in stream)
    print(f"  Raw messages: {raw_count}")
    reader = stream.into_reader()

    stream = reader.messages()
    msg_count = sum(1 for _ in stream)
    print(f"  Messages: {msg_count}")


if __name__ == "__main__":
    main()

File: examples/typescript/multiple_streams.ts
import { Reader } from "@mcapable/wasm";

import { sampleBytes } from "./common";

function main(): void {
  let reader = Reader.fromBytes(sampleBytes());

  console.log("Chunk analysis:");
  let chunkStream = reader.chunks();
  let chunkCount = 0;
  let totalUncompressed = 0;
  for (;;) {
    const item = chunkStream.next();
    if (!item) break;
    const chunk = item as { uncompressed_size: number };
    chunkCount += 1;
    totalUncompressed += chunk.uncompressed_size;
  }
  console.log(`  ${chunkCount} chunks`);
  console.log(`  ${totalUncompressed} bytes uncompressed total`);
  reader = chunkStream.into_reader();

  console.log("\nMessage counts per channel:");
  let msgStream = reader.messages();
  const counts = new Map<number, number>();
  for (;;) {
    const item = msgStream.next();
    if (!item) break;
    const msg = item as { channel_id: number };
    counts.set(msg.channel_id, (counts.get(msg.channel_id) ?? 0) + 1);
  }
  for (const [channelId, count] of counts.entries()) {
    console.log(`  channel ${channelId}: ${count} messages`);
  }
  reader = msgStream.into_reader();

  console.log("\nTime range analysis:");
  let rawStream = reader.raw_messages();
  let minTime: number | null = null;
  let maxTime: number | null = null;
  for (;;) {
    const item = rawStream.next();
    if (!item) break;
    const msg = item as { log_time: number };
    minTime = minTime === null ? msg.log_time : Math.min(minTime, msg.log_time);
    maxTime = maxTime === null ? msg.log_time : Math.max(maxTime, msg.log_time);
  }
  if (minTime !== null && maxTime !== null) {
    console.log(`  start: ${minTime}`);
    console.log(`  end:   ${maxTime}`);
    console.log(`  duration: ${maxTime - minTime}`);
  }
  reader = rawStream.into_reader();

  console.log("\nStream type comparison:");
  let recordStream = reader.records();
  let recordCount = 0;
  while (recordStream.next()) {
    recordCount += 1;
  }
  console.log(`  Records: ${recordCount}`);
  reader = recordStream.into_reader();

  chunkStream = reader.chunks();
  chunkCount = 0;
  while (chunkStream.next()) {
    chunkCount += 1;
  }
  console.log(`  Chunks: ${chunkCount}`);
  reader = chunkStream.into_reader();

  rawStream = reader.raw_messages();
  let rawCount = 0;
  while (rawStream.next()) {
    rawCount += 1;
  }
  console.log(`  Raw messages: ${rawCount}`);
  reader = rawStream.into_reader();

  msgStream = reader.messages();
  let msgCount = 0;
  while (msgStream.next()) {
    msgCount += 1;
  }
  console.log(`  Messages: ${msgCount}`);
}

main();
